File: Study Plan/Top Interview 150/228 Summary Ranges/SummaryRanges.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    228. Summary Ranges
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/summary-ranges/solutions/4426207/clean-easy-to-follow-solution-in-c-beats-83-in-runtime/
    
    - Runtime 135 ms (Beats 83.92% of users with C#)
    - Memory 43.52 MB (Beats 14.87% of users with C#)
*/

public class Solution
{
    public IList<string> SummaryRanges(int[] nums)
    {
        var ranges = new List<string>();

        if (nums.Length == 0)
        {
            return ranges;
        }

        var index = 0;

        // Cycle the entire collection
        while (index < nums.Length)
        {
            var a = nums[index];
            var b = a;

            // Loop while a + 1 = b to define the range
            // (and index is within bounds)
            while (index < nums.Length - 1
                && nums[index] + 1 == nums[index + 1])
            {
                b = nums[index + 1];
                index++;
            }

            // There is more than one element in the range
            if (a != b)
            {
                ranges.Add(string.Format("{0}->{1}", a, b));
            }
            // Just a single element
            else
            {
                ranges.Add(string.Format("{0}", a));
            }

            index++;
        }

        return ranges;
    }
}

File: Study Plan/Top Interview 150/0228 Summary Ranges/SummaryRanges.java
/*  
    Problem
    --------------------------------------------------------------------------------
    228. Summary Ranges
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/summary-ranges/solutions/4453926/java-time-and-space-complexity-o-n-beats-90-in-runtime/
    - Runtime 2 ms (Beats 90.76% of users with Java)
    - Memory 41.40 MB (Beats 9.42% of users with Java)
*/

class Solution {
    public List<String> summaryRanges(int[] nums) {
        ArrayList<String> ranges = new ArrayList<>();
        if (nums.length == 0) {
            return ranges;
        }
        int index = 0;
        // Cycle the entire collection
        while (index < nums.length) {
            int a = nums[index];
            int b = a;
            // Loop while a + 1 = b to define the range
            // (and index is within bounds)
            while (index < nums.length - 1
                && nums[index] + 1 == nums[index + 1]) {
                b = nums[index + 1];
                index++;
            }
            // There is more than one element in the range
            if (a != b) {
                ranges.add(String.format("%s->%s", a, b));
            // Just a single element
            } else {
                ranges.add(String.format("%s", a));
            }
            index++;
        }
        return ranges;
    }
}